syncer/syncer.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	appstudioshared "github.com/redhat-appstudio/managed-gitops/appstudio-shared/apis/appstudio.redhat.com/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Syncer struct {
+	client client.Client
+	ctx    context.Context
+	logger logr.Logger
+}
+
+// NewSyncer creates a new Syncer with the given client and logger.
+func NewSyncer(client client.Client, logger logr.Logger) *Syncer {
+	return NewSyncerWithContext(client, logger, context.TODO())
+}
+
+// NewSyncerWithContext creates a new Syncer with the given client, logger and context.
+func NewSyncerWithContext(client client.Client, logger logr.Logger, ctx context.Context) *Syncer {
+	return &Syncer{
+		client: client,
+		ctx:    ctx,
+		logger: logger,
+	}
+}
+
+// SetContext sets a new context for the Syncer.
+func (s *Syncer) SetContext(ctx context.Context) {
+	s.ctx = ctx
+}
+
+// SyncSnapshot syncs a Snapshot into the given namespace. If an exiting one is found, no operations will be taken.
+func (s *Syncer) SyncSnapshot(snapshot *appstudioshared.ApplicationSnapshot, namespace string) error {
```

Comment:
why not just have this function be in release_adapter.go and not go through all the overhead of creating a new syncer?

Reply:
We are going to need to sync more stuff (Components and maybe Applications data) and that will be done here. Now, I'm not sold about the struct itself and I wanted to go through the `loader.go` design before deciding whether to use methods or functions.

Reply:
Okay so I understand it as the design isn't decided on yet then and we will revisit? I am fine with holding off complaining until `loader.go` is designed if there is a ticket for `loader.go`

Reply:
Yeah, I'd say this syncer is a very beta thing. I'm not happy at all with the way in which snapshots are being synced right now although I'm all-in with having a `syncer.go` file to take care of all the syncing we have to do. Regarding the `loader.go` file, there's no ticket yet, but there will be one.